types.d.ts

```diff
@@ -0,0 +1,241 @@
+ *
+ * @example
+ * ```javascript
+ * import { v4 as uuidv4 } from 'uuid';
+ *
+ * uuidv4(); // ⇨ '1b9d6bcd-bbfd-4b2d-9b5d-ab8dfbbd4bed'
+ * ```
+ *
+ * @example
+ *
+ * ```javascript
+ * import { v4 as uuidv4 } from 'uuid';
+ *
+ * const v4options = {
+ *   random: [
+ *     0x10, 0x91, 0x56, 0xbe, 0xc4, 0xfb, 0xc1, 0xea, 0x71, 0xb4, 0xef, 0xe1, 0x67, 0x1c, 0x58, 0x36,
+ *   ],
+ * };
+ * uuidv4(v4options); // ⇨ '109156be-c4fb-41ea-b1b4-efe1671c5836'
+ * ```
+ */
+export function v4(options?: V4Options | null): string
+export function v4<T extends ArrayLike<number>>(options: V4Options | null | undefined, buffer: T, offset?: number): T
+
+/**
+ * Create an RFC version 5 (namespace w/ SHA-1) UUID
+ *
+ * @param name
+ * @param namespace  Namespace UUID
+ * @param buffer If specified, uuid will be written here in byte-form, starting at `offset`
+ * @param offset Index to start writing UUID bytes in `buffer`
+ * @returns UUID `String` if no `buffer` is specified, otherwise returns `buffer`
+ *
+ * @note The RFC `DNS` and `URL` namespaces are available as `v5.DNS` and `v5.URL`
+ *
+ * @example
+ * ```javascript
+ * import { v5 as uuidv5 } from 'uuid';
+ *
+ * // Define a custom namespace.  Readers, create your own using something like
+ * // https://www.uuidgenerator.net/
+ * const MY_NAMESPACE = '1b671a64-40d5-491e-99b0-da01ff1f3341';
+ *
+ * uuidv5('Hello, World!', MY_NAMESPACE); // ⇨ '630eb68f-e0fa-5ecc-887a-7c7a62614681'
+ * ```
+ *
+ * @example
+ * ```javascript
+ * import { v5 as uuidv5 } from 'uuid';
+ *
+ * uuidv5('https://www.w3.org/', uuidv5.URL); // ⇨ 'c106a26a-21bb-5538-8bf2-57095d1976c1'
+ * ```
+ */
+export function v5(name: string | ArrayLike<number>, namespace: string | ArrayLike<number>): string
+export function v5<T extends ArrayLike<number>>(name: string | ArrayLike<number>, namespace: string | ArrayLike<number>, buffer: T, offset?: number): T
+export namespace v5 { export const DNS: string; export const URL: string; }
+
+/**
+ * Test a string to see if it is a valid UUID
+ *
+ * @param str `String` to validate
+ * @returns `true` if string is a valid UUID, `false` otherwise
+ *
+ * @example
+ * ```javascript
+ * import { validate as uuidValidate } from 'uuid';
+ *
+ * uuidValidate('not a UUID'); // ⇨ false
+ * uuidValidate('6ec0bd7f-11c0-43da-975e-2a8ad9ebae0b'); // ⇨ true
+ * ```
+ *
+ * @example
+ * Using `validate` and `version` together it is possible to do per-version validation, e.g. validate for only v4 UUIds.
+ *
+ * ```javascript
+ * import { version as uuidVersion } from 'uuid';
+ * import { validate as uuidValidate } from 'uuid';
+ *
+ * function uuidValidateV4(uuid) {
+ *   return uuidValidate(uuid) && uuidVersion(uuid) === 4;
+ * }
+ *
+ * const v1Uuid = 'd9428888-122b-11e1-b85c-61cd3cbb3210';
+ * const v4Uuid = '109156be-c4fb-41ea-b1b4-efe1671c5836';
+ *
+ * uuidValidateV4(v4Uuid); // ⇨ true
+ * uuidValidateV4(v1Uuid); // ⇨ false
+ * ```
+ */
+export function validate(str: string): boolean
```

Comment:
In order to support #606:

```suggestion
export function validate(str: any): str is string
```

Reply:
nominal types would have made this way cleaner 😛

Reply:
Yeah for sure!